cpu/native/socket_zep/socket_zep.c

```diff
@@ -408,58 +222,351 @@ static void _send_zep_hello(socket_zep_t *dev)
+    case IEEE802154_AF_PAN_COORD:
+        return -ENOTSUP;
+    }
+
+    return 0;
+}
+
+static int _config_src_addr_match(ieee802154_dev_t *dev, ieee802154_src_match_t cmd, const void *value)
+{
+    (void) dev;
+    (void) cmd;
+    (void) value;
+    return -ENOTSUP;
+}
+
+static int _set_frame_filter_mode(ieee802154_dev_t *dev, ieee802154_filter_mode_t mode)
+{
+    socket_zep_t *zepdev = dev->priv;
+    zepdev->filter_mode = mode;
+    return 0;
+}
+
+static int _request_set_trx_state(ieee802154_dev_t *dev, ieee802154_trx_state_t state)
+{
+    (void) dev;
+    (void) state;
+    return 0;
+}
+
+static int _confirm_set_trx_state(ieee802154_dev_t *dev)
+{
+    (void) dev;
+    return 0;
+}
+
+static int _write(ieee802154_dev_t *dev, const iolist_t *iolist)
+{
+    socket_zep_t *zepdev = dev->priv;
+    unsigned n = iolist_count(iolist);
+    size_t bytes = iolist_size(iolist) + sizeof(uint16_t); /* FCS field */;
+    uint8_t *out = zepdev->snd_buf;
+    uint16_t chksum = 0;
+
+    out += _zep_hdr_fill(zepdev, (void *)out, bytes);
+
+    /* make sure we are not overflowing the TX buffer */
+    if (out + bytes > zepdev->snd_buf + sizeof(zepdev->snd_buf)) {
+        return -ENOBUFS;
+    }
+
+    for (unsigned i = 0; i < n; i++) {
+        memcpy(out, iolist->iol_base, iolist->iol_len);
+        chksum = ucrc16_calc_le(iolist->iol_base, iolist->iol_len,
+                                UCRC16_CCITT_POLY_LE, chksum);
+        out += iolist->iol_len;
+        iolist = iolist->iol_next;
+    }
+    chksum = byteorder_htols(chksum).u16;
+    memcpy(out, &chksum, sizeof(chksum));
+
+    zepdev->snd_len = (out + 2) - zepdev->snd_buf;
+
+    return bytes;
+}
+
+static int _request_transmit(ieee802154_dev_t *dev)
+{
+    socket_zep_t *zepdev = dev->priv;
+
+    dev->cb(dev, IEEE802154_RADIO_INDICATION_TX_START);
+
+    int res = real_write(zepdev->sock_fd, zepdev->snd_buf, zepdev->snd_len);
+
+    dev->cb(dev, IEEE802154_RADIO_CONFIRM_TX_DONE);
+
+    if (res < 0) {
+        return res;
+    }
+
+    return 0;
+}
+
+static int _confirm_transmit(ieee802154_dev_t *dev, ieee802154_tx_info_t *info)
+{
+    (void) dev;
+
+    /* TODO: handle ACK */
+    if (info) {
+        info->status = TX_STATUS_SUCCESS;
+    }
+
+    return 0;
+}
+
+int _len(ieee802154_dev_t *dev)
+{
+    size_t size;
+    socket_zep_t *zepdev = dev->priv;
+
+    int res = real_ioctl(zepdev->sock_fd, FIONREAD, &size);
+    if (res < 0) {
+        DEBUG("socket_zep::len: error reading FIONREAD: %s", strerror(errno));
+        return res;
+    }
+
+    return size;
+}
+
+static int _read(ieee802154_dev_t *dev, void *buf, size_t max_size,
+                          ieee802154_rx_info_t *info)
+{
+    int res;
+    socket_zep_t *zepdev = dev->priv;
+
+    DEBUG("socket_zep::read: reading up to %u bytes into %p\n", max_size, buf);
+
+    res = real_read(zepdev->sock_fd, buf, max_size);
+
+    DEBUG("socket_zep::read: got %d bytes\n", res);
+
+    if (res <= 0) {
+        return res;
```

Comment:
I think you could either return 0 (since the frame is not written into `buf`) or we could add return codes to the `read` function. As it is now, it only returns the number of bytes or `-ENOBUFS`